README.md
Cerner Tomcat Cookbook
======================

[![Build Status](https://travis-ci.org/cerner/cerner_kafka.svg?branch=master)](https://travis-ci.org/cerner/cerner_tomcat)
[![Cookbook Version](https://img.shields.io/cookbook/v/cerner_tomcat.svg)](https://community.opscode.com/cookbooks/cerner_tomcat)

A Chef cookbook to install [Apache Tomcat](http://tomcat.apache.org/).

View the [Change Log](CHANGELOG.md) to see what has changed.

Requirements
------------

### Platforms

 * Ubuntu
 * CentOS

### Chef

 * Chef 12+

### Cookbooks

 * java
 * ulimit
 * logrotate
 * poise
 * poise-archive

Usage
-----

The cookbook provides a resource for you to use to install and configure tomcat.

The `cerner_tomcat` resource has support for `remote_file`, `cookbook_file`, and `template` as sub-resources of the cerner_tomcat resource. When using these sub-resources, the parent `cerner_tomcat` resource's path, user, group and sensitive properties are provided for free. These sub-resources can also be called from outside the `cerner_tomcat` block, if desired, by prepending `cerner_tomcat_` to the resource call and provided a `parent` property.

Example:
```ruby
cerner_tomcat 'my_tomcat' do
  # resource config
end

cerner_tomcat_remote_file do
   parent 'my_tomcat'
   source '...'
end
```

`cerner_tomcat` delays restart, when necessary, for the end of the main converge. This is in order to avoid restarting the service multiple times during a chef-client run. If any service dependencies are updated and require a service restart, this can be done by notifying the resource with a restart action.

Example:
```ruby
cerner_tomcat 'my_tomcat' do
  # resource config
end

dependent_resource 'dep' do
  # resource config
  notifies :restart, 'cerner_tomcat[my_tomcat]'
end
```

### cerner_tomcat

A resource for installing and configuring tomcat.

Actions: `:enable`, `:disable`, `:start`, `:stop`, `:restart`, `:reload` (default `:enable`)

Parameters:
 * `instance_name`: The name of the tomcat instance. (default = name of resource)
 * `user`: The user that will own and run the tomcat process (default = `tomcat`)
 * `group`: The group the tomcat user will be part of (default = `tomcat`)
 * `base_dir`: The path to the directory where the tomcat instance will be contained (default = `/opt/tomcat`)
 * `log_dir`: The path to the directory where the tomcat logs will be (default = `/var/log/tomcat`)
 * `log_rotate_options`: A hash or ruby block of options to configure log rotate. These will be merged with and override the defaults provided (view `libraries/cerner_tomcat.rb` for defaults)
 * `version`: The version of tomcat to install. This effectively builds the tomcat_url from a known URL (`repo1.maven.org`) with the given version (default=`8.0.21`)
 * `tomcat_url`: The URL to the tomcat binary used to install tomcat. This will override the url provided by `version`. NOTE: `version` needs to still be up to date in order to install tomcat properly
 * `shutdown_timeout`: The timeout used when trying to shutdown the tomcat service (default = `60`). If the timeout is reached, diagnostics are captured about the instance
 and the a force shutdown is applied.
 * `java_settings`: A Hash of java settings to be applied to the tomcat process (default = `{}`)
 * `env_vars`: A Hash of environment variables to be available when starting tomcat (default = `{}`)
 * `init_info`: A Hash or ruby block of options to configure the init script. These will be merged with and override the defaults provided (view `libraries/cerner_tomcat.rb` for defaults)
 * `install_java`: A boolean that indicates if we should try to install java (default=`true`)
 * `limits`: A Hash or ruby block of limits applied to the owner user of the tomcat process (default=`{ 'open_files' => 32_768, 'max_processes' => 1024 }`)
 * `create_user`: A boolean that indicates if the service user should be created (default=`true`)
 * `sensitive`: A boolean used to ensure sensitive resource data is not logged by the chef-client. Remote_file, cookbook_file, and template sub-resources can be overriden (default=true)
 * `service_manager`: The service management framework to use. Supported frameworks include `:sysvinit` and `:upstart`. If using `:upstart` then init_info, shutdown_timeout, and any provided health_checks will be ignored (default=`:sysvinit`)
 * `download_dir`: The path to the directory where the tomcat package will be download. user/group should have permission to read (default = `/var/cerner_tomcat`)
 * `user_shell`: The shell used by the service user for tomcat. If not supplied will fall back to poise_service_user defaults `[/bin/nologin, /usr/bin/nologin, /bin/false]`

Example:
``` ruby
cerner_tomcat "my_tomcat" do
  user "my_user"
  group "my_group"
  base_dir "/opt/my_dir"
  log_dir "/var/log/my_dir"
  log_rotate_options do
    frequency: 'daily'
    size: '10M'
    maxsize: '100M'
  end
  version "7.0.49"
  shutdown_timeout 120
  java_settings("-Xms" => "512m",
                 "-Xmx" => "512m",
                 "-XX:PermSize=" => "384m",
                 "-XX:MaxPermSize=" => "384m"
                 "-XX:+UseParNewGC" => "")
  env_vars("MY_VAR" => "MY_VALUE")
  init_info('Default-Start' => '1 2 3 4 5')
  install_java false
  limits(
    'open_files' => 32_768,
    'max_processes' => 1024
  )
end
```

Sub-Resource Parameters:

#### health_check

A health check to be run when starting the tomcat service which uses HTTP to hit a web app's
health resource to verify it started correctly and is working. The name of the resource
should be the URL to your web app's health resource.

Parameters:

 * `http_method`: The HTTP method to use when running the health check (default=`GET`)
 * `backoff`: An array that indicates how back off should be handled, including the number of times and how long to wait before trying again (default=`[0, 5, 10, 30, 30, 60]`)
 * `time_bound`: The amount of time to wait before timing out a request (default=`3`)
 * `args`: Additional command line arguments to pass into cURL
 * `redirects_healthy`: Count a redirect response (300-399) as healthy

Example:
``` ruby
cerner_tomcat "my_tomcat" do
  # Uses defaults
  health_check "http://localhost:8080/myApp/health"

  health_check "http://localhost:8080/myApp/health" do
    http_method 'POST'
    backoff [0, 1, 2, 3]
    time_bound 5
  end
end
```

#### web_app

A web application to be included in the tomcat installation. The name of the sub-resource should
be the context root to be used for the application. This sub-resource also accepts `cookbook_file`,
`remote_file`, and `template` sub-resources just like those listed above.

Parameters:
 * `source`: The URL to the war file of the application (required)
 * `checksum`: The SHA256 checksum of the war file of the application (optional)

Example:
``` ruby
cerner_tomcat "my_tomcat" do
  web_app "myApp" do
    source "http://myRepo.com/myApp-1.0.war"
    checksum "feccf30cad8aa5aea7a4e661ca9c32bab44d3df8ef5d97bd9c7eacf5c841bb17"

    cookbook_file "WEB-INF/classes/app.properties" do
      source "app.properties"
    end

    remote_file "WEB-INF/lib/mysql.jar" do
      source "http://repo1.maven.org/maven2/mysql/mysql-connector-java/5.1.25/mysql-connector-java-5.1.25.jar"
    end

    template "WEB-INF/classes/log4j.properties" do
      source "log4j.properties.erb"
    end
  end
end
```

### Troubleshooting

The cookbook additionally configures a `diagnostic` command option as part of the `sysvinit` installation of the service. This command
captures various metrics of the JVM and bundles it in a `tar.gz` within a temp directory for later evaluation. Diagnostics which are
included in this bundle:

* JVM performance counters: Series of stats on the JVM, which are a lot of simple facts of the run-time at a low cost of acquiring
(filename: `<service_name>_perfcount.log`).
* JVM thread dump: Helpful to identify non-daemon threads and JVM information about each thread with their native thread ID
(filename: `<service_name>_thread_dump.log`).
* Native thread logs: Helpful to correlate CPU utilization to native threads which are tied to in the JVM thread dump with the
native thread ID (nid). These are samplings from `top` which include native thread utilization (filename: `<service_name>_native_thread.log`).
* JVM GC class histogram: Helpful if JVM seems hung, and identified GC threads are utilizing high CPU, to then evaluate the amount
of objects / types being created (filename: `<service_name>_gc_class_histogram.log`).

If using the `sysvinit` installation of the service, and the Tomcat service does not shutdown in a timely manner (within the defined
timeout period), it will include a thread dump as part of the service stdout (ex. `catalina.out`), and may additionally invoke this
diagnostic bundle to be created before forcing the service process to shutdown (through an OS signal). This can be helpful to further
evaluate if a service is consistently requiring a forceful shutdown (web application may have a non-daemon thread not being shutdown).

Contributing
------------

This project is licensed under the Apache License, Version 2.0.

When contributing to the project please add your name to the CONTRIBUTORS.txt file. Adding your name to the CONTRIBUTORS.txt file
signifies agreement to all rights and reservations provided by the License.

To contribute to the project execute a pull request through github. The pull request will be reviewed by the community and merged
by the project committers. Please attempt to conform to the test, code conventions, and code formatting standards if any
are specified by the project before submitting a pull request.

Releases
--------

Releases should happen regularly after most changes. Feel free to request a release by logging an issue.

Committers
----------

For information related to being a committer check [here](COMMITTERS.md).

LICENSE
-------

Copyright 2013 Cerner Innovation, Inc.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

[http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0) Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
Change Log
==========

[3.7.0 - 08-18-2021](https://github.com/cerner/cerner_tomcat/issues?milestone=16&state=closed)
----------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-67](https://github.com/cerner/cerner_tomcat/issues/67) : Add support for providing a user shell to tomcat's service user

[3.6.0 - 01-15-2020](https://github.com/cerner/cerner_tomcat/issues?milestone=15&state=closed)
----------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-66](https://github.com/cerner/cerner_tomcat/issues/66) : Changed default tomcat URL to https

[3.5.0 - 06-25-2019](https://github.com/cerner/cerner_tomcat/issues?milestone=14&state=closed)
----------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-64](https://github.com/cerner/cerner_tomcat/issues/64) : updating limit version information

[3.4.1 - 03-11-2019](https://github.com/cerner/cerner_tomcat/issues?milestone=13&state=closed)
----------------------------------------------------------------------------------------------

  * [Bug] [Issue-63](https://github.com/cerner/cerner_tomcat/issues/63) : Adding redirects_healthy return value

[3.4.0 - 03-01-2019](https://github.com/cerner/cerner_tomcat/issues?milestone=12&state=closed)
----------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-61](https://github.com/cerner/cerner_tomcat/issues/61) : Add option to count redirects as healthy
  * [Enhancement] [Issue-60](https://github.com/cerner/cerner_tomcat/issues/60) : Use download_dir for diagnostics
  * [Enhancement] [Issue-58](https://github.com/cerner/cerner_tomcat/issues/58) : Install requires opening permissions on /var/chef/cache

[3.3.0 - 02-20-2019](https://github.com/cerner/cerner_tomcat/issues?milestone=11&state=closed)
----------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-57](https://github.com/cerner/cerner_tomcat/issues/57) : update to support java 2.x

[3.2.0 - 01-02-2018](https://github.com/cerner/cerner_tomcat/issues?milestone=9&state=closed)
---------------------------------------------------------------------------------------------

  * [Feature] [Issue-52](https://github.com/cerner/cerner_tomcat/issues/52) : Accept checksum for web_app source
  * [Enhancement] [Issue-46](https://github.com/cerner/cerner_tomcat/issues/46) : Fixed foodcritic issues after Chef 13 release

[3.1.0 - 03-06-2017](https://github.com/cerner/cerner_tomcat/issues?milestone=7&state=closed)
---------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-41](https://github.com/cerner/cerner_tomcat/issues/41) : Added support for setting the nofile ulimit in the upstart configurat…
  * [Feature] [Issue-5](https://github.com/cerner/cerner_tomcat/issues/5) : Kill process when it doesn't stop cleanly and gather information

[3.0.0 - 01-16-2017](https://github.com/cerner/cerner_tomcat/issues?milestone=6&state=closed)
---------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-22](https://github.com/cerner/cerner_tomcat/issues/22) : Upgrade off init to upstart/systemd/etc
  * [Bug] [Issue-21](https://github.com/cerner/cerner_tomcat/issues/21) : find_pid within init.d script can qualify different Tomcat installations 
  * [Enhancement] [Issue-4](https://github.com/cerner/cerner_tomcat/issues/4) : Sub resources should provide all options (remote_file, cookbook_file, template)

[2.5.0 - 01-02-2018](https://github.com/cerner/cerner_tomcat/issues?milestone=10&state=closed)
----------------------------------------------------------------------------------------------

  * [Feature] [Issue-53](https://github.com/cerner/cerner_tomcat/issues/53) : Accept checksum for web_app source (2.x patchback)

[2.4.0 - 12-11-2017](https://github.com/cerner/cerner_tomcat/issues?milestone=8&state=closed)
---------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-47](https://github.com/cerner/cerner_tomcat/issues/47) : Fix foodcritic 2x
  * [Enhancement] [Issue-45](https://github.com/cerner/cerner_tomcat/issues/45) : Kill process when it doesn't stop cleanly and gather information (2.x)

[2.3.0 - 10-26-2016](https://github.com/cerner/cerner_tomcat/issues?milestone=4&state=closed)
---------------------------------------------------------------------------------------------

  * [Bug] [Issue-26](https://github.com/cerner/cerner_tomcat/issues/26) : Install block has unnecessary not_if guard
  * [Enhancement] [Issue-25](https://github.com/cerner/cerner_tomcat/issues/25) : Adding additional properties to sub resources
  * [Enhancement] [Issue-23](https://github.com/cerner/cerner_tomcat/issues/23) : Support additional command line options for health checks
  * [Enhancement] [Issue-15](https://github.com/cerner/cerner_tomcat/issues/15) : Add a flag to disable user/group creation

[2.2.0 - 09-13-2016](https://github.com/cerner/cerner_tomcat/issues?milestone=3&state=closed)
---------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-17](https://github.com/cerner/cerner_tomcat/issues/17) : ulimit settings are hard-coded in the provider
  * [Enhancement] [Issue-16](https://github.com/cerner/cerner_tomcat/issues/16) : Reduce the number of service start-up events on a single chef-client execution

[2.1.1 - 05-02-2016](https://github.com/cerner/cerner_tomcat/issues?milestone=2&state=closed)
---------------------------------------------------------------------------------------------

  * [Bug] [Issue-9](https://github.com/cerner/cerner_tomcat/issues/9) : Restart service on java setting changes

libraries/cerner_tomcat.rb
#
# Copyright 2013 Cerner Innovation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'poise'
require 'poise_service'

module CernerTomcat
  module Resources
    module Tomcat
      class Resource < Chef::Resource
        include Poise(container: true, container_namespace: false)
        provides(:cerner_tomcat)
        include PoiseService::ServiceMixin

        def initialize(*args)
          super

          @health_checks = []

          # so our default below can work.
          remove_instance_variable(:@sensitive) if instance_variable_defined?(:@sensitive)

          _rewire_sub_resources! if node
        end

        attr_reader :health_checks

        def health_check(uri, &block)
          health_check_block = ::CernerTomcat::HealthCheckBlock.new(uri)
          health_check_block.evaluate(&block)
          @health_checks << health_check_block
        end

        attribute :instance_name, kind_of: String, name_attribute: true
        attribute :user, kind_of: String, default: 'tomcat'
        attribute :group, kind_of: String, default: 'tomcat'
        attribute :base_dir, kind_of: String, default: '/opt/tomcat'
        attribute :log_dir, kind_of: String, default: '/var/log/tomcat'
        attribute :log_rotate_options, option_collector: true, default: lazy { default_log_rotate_options }
        attribute :version, kind_of: String, default: '8.0.21'
        attribute :tomcat_url, kind_of: String, default: lazy { default_tomcat_url }
        attribute :shutdown_timeout, kind_of: Integer, default: 60
        attribute :java_settings, kind_of: Hash, default: {}
        attribute :env_vars, kind_of: Hash, default: {}
        attribute :init_info, option_collector: true, default: lazy { default_init_info }
        attribute :install_java, equal_to: [true, false], default: true
        attribute :limits, option_collector: true, default: { 'open_files' => 32_768, 'max_processes' => 1024 }
        attribute :create_user, equal_to: [true, false], default: false
        attribute :sensitive, equal_to: [true, false], default: true
        attribute :service_manager, equal_to: [:sysvinit, :upstart], default: :sysvinit
        attribute :download_dir, kind_of: String, default: '/var/cerner_tomcat'
        attribute :user_shell, kind_of: String, default: nil

        def install_dir
          ::File.join(base_dir, instance_name)
        end

        def command
          "#{install_dir}/bin/catalina.sh run"
        end

        def default_log_rotate_options
          {
            'path' => [],
            'frequency' => 'daily',
            'rotate' => 3,
            'options' => %w(notifempty copytruncate compress),
            'size' => '10M',
            'maxsize' => '100M'
          }
        end

        def default_init_info
          {
            'Default-Start' => '3 4 5',
            'Default-Stop' => '0 1 2 6',
            'Required-Start' => '',
            'Required-Stop' => '',
            'Should-Start' => '',
            'Should-Stop' => ''
          }
        end

        def default_tomcat_url
          "https://repo1.maven.org/maven2/org/apache/tomcat/tomcat/#{version}/tomcat-#{version}.tar.gz"
        end

        private

        # Rewire our cerner_tomcat sub-resources so they dont require the cerner_tomcat_ prefix
        def _rewire_sub_resources!
          rewire_map = {
            'remote_file' => 'cerner_tomcat_remote_file',
            'template' => 'cerner_tomcat_template',
            'cookbook_file' => 'cerner_tomcat_cookbook_file',
            'web_app' => 'cerner_tomcat_web_app'
          }
          # Generate stub methods for all the rewiring.
          rewire_map.each do |new_name, old_name|
            # This is defined as a singleton method on self so it looks like
            # the DSL but is scoped to just this context.
            define_singleton_method(new_name) do |name=nil, *args, &block|
              # Store the caller to correct the source_line.
              created_at = caller[0]
              public_send(old_name, name, *args) do
                # Set the declared type to be the native name.
                self.declared_type = self.class.resource_name
                # Fix the source location. for chef 12.4 we could do this with the
                # declared_at parameter on the initial send.
                self.source_line = created_at
                # Run the original block.
                instance_exec(&block) if block
              end
            end
          end
        end
      end

      class Provider < Chef::Provider
        include Poise
        provides(:cerner_tomcat)
        include PoiseService::ServiceMixin

        def action_enable
          notifying_block do
            log_dir = ::File.join(new_resource.log_dir, new_resource.instance_name)
            tomcat_url = new_resource.tomcat_url
            tomcat_file = "cerner_tomcat_#{new_resource.instance_name}-apache-tomcat.tar.gz"

            # Install java
            run_context.include_recipe 'java' if new_resource.install_java

            # Ensure curl is installed if we have any health checks
            package "curl_#{new_resource.instance_name}" do
              action :install
              package_name 'curl'
              not_if { new_resource.health_checks.empty? }
            end

            # Setup user/group
            poise_service_user "#{new_resource.instance_name} user" do
              user new_resource.user
              group new_resource.group
              only_if { new_resource.create_user }
              shell new_resource.user_shell unless new_resource.user_shell.nil?
            end

            ulimits = new_resource.limits
            user_ulimit new_resource.user do
              filehandle_limit ulimits['open_files']
              process_limit ulimits['max_processes']
            end

            # From the requirements of the cookbook:
            # If you're on Ubuntu, you'll also need to add recipe[ulimit] to your
            # runlist, or the files created by this cookbook will be ignored
            run_context.include_recipe 'ulimit' if platform_family?('ubuntu')

            # Create base directory
            directory new_resource.base_dir do
              action :create
              owner new_resource.user
              group new_resource.group
              mode '0755'
              recursive true
            end

            directory new_resource.download_dir do
              owner new_resource.user
              group new_resource.group
              mode '0755'
            end

            remote_file "#{new_resource.download_dir}/#{tomcat_file}" do
              source tomcat_url
              owner new_resource.user
              group new_resource.group
              mode '0755'
              backup false
              notifies :stop, new_resource, :immediately
              notifies :unpack, "poise_archive[#{new_resource.download_dir}/#{tomcat_file}]", :immediately
            end

            web_app_dir = ::File.join(new_resource.install_dir, 'webapps')

            # Install tomcat
            poise_archive "#{new_resource.download_dir}/#{tomcat_file}" do
              action :nothing
              destination new_resource.install_dir
              user new_resource.user
              group new_resource.group
              notifies :delete, "directory[#{web_app_dir}]", :immediately
              notifies :restart, new_resource
            end

            directory web_app_dir do
              owner new_resource.user
              group new_resource.group
              mode '0755'
              recursive true
            end

            directory new_resource.log_dir do
              owner new_resource.user
              group new_resource.group
              mode '0755'
              recursive true
            end

            # Link tomcat log files into /var/log/tomcat/[instanceName]
            link log_dir do
              to "#{new_resource.install_dir}/logs"
              owner new_resource.user
              group new_resource.group
              mode '0755'
            end

            # Ensure the right paths are always added
            log_rotate_options = new_resource.log_rotate_options
            log_rotate_options['path'] += ["#{log_dir}/*.out", "#{log_dir}/*.log"]

            # Add log rotate config for tomcat
            logrotate_app "tomcat-#{new_resource.instance_name}" do
              log_rotate_options.each do |k, v|
                send k, v
              end
            end

            # Configure environment vars and java settings
            template "#{new_resource.install_dir}/bin/setenv.sh" do
              source 'setenv.sh.erb'
              cookbook 'cerner_tomcat'
              mode '0755'
              backup false
              sensitive new_resource.sensitive
              owner new_resource.user
              group new_resource.group
              variables(
                env_vars: new_resource.env_vars,
                java_settings: new_resource.java_settings
              )
              notifies :restart, new_resource
            end
          end
          super
        end

        def action_disable
          # Since we are recursively deleting directories guard against empty string which
          # could result in deleting more than what we want
          fail 'base_dir cannot be empty string' if new_resource.base_dir.empty?
          fail 'instance_name cannot be empty string' if new_resource.instance_name.empty?
          fail 'log_dir cannot be empty string' if new_resource.log_dir.empty?

          super
          notifying_block do
            directory "#{new_resource.base_dir}/#{new_resource.instance_name}" do
              action :delete
              recursive true
            end

            file "#{new_resource.log_dir}/#{new_resource.instance_name}" do
              action :delete
            end
          end
        end

        def service_options(service)
          service.command(new_resource.command)
          service.directory(new_resource.install_dir)
          service.user(new_resource.user)
          service.environment( { 'CATALINA_HOME' => new_resource.install_dir } )
          service.options(
            :cerner_tomcat_sysvinit,
            template: 'cerner_tomcat:tomcat_init.sh.erb',
            script_path: "/etc/init.d/tomcat_#{new_resource.instance_name}"
          )
          service.options(
            :cerner_tomcat_upstart,
            template: 'cerner_tomcat:upstart.conf.erb'
          )
          service.provider("cerner_tomcat_#{new_resource.service_manager}".to_sym)
        end
      end
    end
  end
end
